Repeat Toolbar Button

Ben2356 opened the pull request:
Added a repeat toolbar button which will provide repeat functionality if it is supported for the current playlist/queue and by the music application. If the repeat functionality is not supported then the button will be hidden. The button is a toggle between 3 different modes: 

Repeat All
Repeat One 
Repeat Off

The repeat button is only available for the audioHmiState as that layout is the only one that contains the images.zip file with the repeat button icons, the regular toolbarHmiState will not have this new repeat button. To support the old toolbarHmiState the custom actions are available to toggle between the repeat options.

I have also changed the checkmark icon next to the currently playing song name to a music note which seems more appropriate in my opinion.

Repeat button in action: 
![1](https://user-images.githubusercontent.com/11298573/94876251-c07dd700-041c-11eb-8c32-bce97d24af6b.jpg)
![2](https://user-images.githubusercontent.com/11298573/94876263-c4115e00-041c-11eb-9cdb-538350c2e43c.jpg)
![3](https://user-images.githubusercontent.com/11298573/94876268-c70c4e80-041c-11eb-9d99-f3ba70a94e70.jpg)
![4](https://user-images.githubusercontent.com/11298573/94876273-ca073f00-041c-11eb-80d1-d06c67ae8098.jpg)

This issue also references this issue: https://github.com/hufman/AndroidAutoIdrive/issues/111

hufman commented on app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/PlaybackView.kt:
```diff
@@ -197,7 +202,10 @@ class PlaybackView(val state: RHMIState, val controller: MusicController, val ca
+			// setting the actions button icon since the button has a book icon by default
+			customActionButton.getImageModel()?.asImageIdModel()?.imageId = musicImageIDs.ACTIONS
+
+			repeatButton!!.getAction()?.asRAAction()?.rhmiActionCallback = RHMIActionCallback { controller.toggleRepeat() }
```
Please use the null-coalescing `repeatButton?.getAction` style, just in case something accidentally doesn't get initialized earlier.

Ben2356 replied:
Ah not sure why I did !! there instead of ?. , changed that

hufman commented on app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/PlaybackView.kt:
```diff
@@ -366,6 +375,26 @@ class PlaybackView(val state: RHMIState, val controller: MusicController, val ca
+			repeatButton.setVisible(true)
+		} else if (repeatButton != null) {
+			repeatButton.setEnabled(false)
+			repeatButton.setVisible(false)
```
The repeatButton should be left visible and disabled when unavailable, matching the shuffle button, so that the full toolbar of 6 buttons keeps its nice appearance :)

Ben2356 replied:
Good point, removed the setVisible line!

hufman commented on app/src/main/java/me/hufman/androidautoidrive/music/controllers/MusicAppController.kt:
```diff
@@ -42,6 +44,10 @@ interface MusicAppController {
 
 	fun isShuffling(): Boolean
 
+	fun isRepeatingAll(): Boolean
```
I think these two functions should be combined to a single function, such as `getRepeatMode()`. It could return a PlaybackStateCompat.REPEAT* constant int, but it could also return a new Enum to represent the tristate repeat mode. These two separate boolean state flags doesn't encode the same meaning.

Ben2356 replied:
I agree it would be better as a single function, I added a new getRepeatMode method that returns a new RepeatMode enum that way it is decoupled from android, spotify, etc

hufman commented on app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/PlaybackView.kt:
```diff
@@ -284,7 +293,7 @@ class PlaybackView(val state: RHMIState, val controller: MusicController, val ca
 			val playlist = RHMIModel.RaListModel.RHMIListConcrete(10)
 			playlist.addRow(PlaylistItem(false, skipBackEnabled, BMWRemoting.RHMIResourceIdentifier(BMWRemoting.RHMIResourceType.IMAGEID, musicImageIDs.SKIP_BACK), L.MUSIC_SKIP_PREVIOUS))
-			playlist.addRow(PlaylistItem(false, true, BMWRemoting.RHMIResourceIdentifier(BMWRemoting.RHMIResourceType.IMAGEID, musicImageIDs.CHECKMARK), title))
+			playlist.addRow(PlaylistItem(false, true, BMWRemoting.RHMIResourceIdentifier(BMWRemoting.RHMIResourceType.IMAGEID, musicImageIDs.SONG), title))
```
I did the checkmark because Spotify did it, but the song icon makes more sense!

Ben2356 replied:
Yeah I remember the original implementation did that too and always thought it was an odd choice so decided since it was such a simple thing I'd roll it into the repeat mode changes 😃

hufman reviewed (changes requested):
Looks great! Just a few small tweaks, and can you rewrite the commit history to have two commits for the two separate changes: `Changes the song icon`, and `Adds the repeat toolbar button`
Thank you so much!!

Ben2356 commented:
> 
> 
> Looks great! Just a few small tweaks, and can you rewrite the commit history to have two commits for the two separate changes: `Changes the song icon`, and `Adds the repeat toolbar button`
> Thank you so much!!

Rebased onto master and squashed those into two commits! You're welcome always happy to help out!

hufman commented on app/src/test/java/me/hufman/androidautoidrive/TestSpotifyMusicAppController.kt:
```diff
@@ -106,13 +107,6 @@ class TestSpotifyMusicAppController {
 
 	@Test
 	fun testCustomActions() {
-		controller.customAction(controller.CUSTOM_ACTION_TURN_REPEAT_ALL_ON)
```
Would it make sense to leave these in place?

Ben2356 replied:
It does, those should be re-added and I must have forgotten about that when I re-added the custom action repeat item. Thanks for catching that!

hufman commented:
This is great stuff! Thank you so much!

Ben2356 commented:
Reordered and squashed commits down!
